torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+    """
+
+    if dist.get_rank() == src_rank:
+        # Emit data
+        buffer = io.BytesIO()
+        torch.save(obj, buffer)
+        data = bytearray(buffer.getbuffer())
+        length_tensor = torch.LongTensor([len(data)]).to(dist_device)
+        data_send_tensor = torch.ByteTensor(data).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        dist.broadcast(data_send_tensor, src=src_rank, group=group, async_op=False)
+    else:
+        # Fetch from the source
+        length_tensor = torch.LongTensor([0]).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        data_recv_tensor = torch.empty([int(length_tensor.item())], dtype=torch.uint8, device=dist_device)
+        dist.broadcast(data_recv_tensor, src=src_rank, group=group, async_op=False)
+        buffer = io.BytesIO(data_recv_tensor.cpu().numpy())
+        obj = torch.load(buffer, map_location=dist_device)
+    return obj
+
+
+# Credits:  classy_vision/generic/distributed_util.py
+def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+    """
+    Recursively searches lists, tuples, dicts and copies tensors to device if
+    possible. Non-tensor values are passed as-is in the result.
+
+    NOTE:  These are all copies, so if there are two objects that reference
+    the same object, then after this call, there will be two different objects
+    referenced on the device.
+    """
+
+    if isinstance(value, torch.Tensor):
+        return value.to(device, non_blocking=non_blocking)
+
+    if isinstance(value, (list, tuple)):
+        values = [_recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for val in value]
+        return values if isinstance(value, list) else tuple(values)
+
+    if isinstance(value, container_abcs.Mapping):
+        return {
+            key: _recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for key, val in value.items()
+        }
+
+    return value
+
+
+def _get_global_rank(group: Any, rank: int) -> int:
+    if group is dist.group.WORLD:
+        return rank
+    else:
+        global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+    return global_rank
+
+
+class ZeROptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ShardedOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters
+    at each rank. Each parameter belongs to a single rank and
+    is not divided among rank.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
```

Comment:
Shall we add a warning block highlighting that this is a prototype feature and subject to change?

Reply:
Ideally I would not change that too much, we've been using it in some research groups for a while, it "just works". I think that some of the interface could change but if possible I would do that during this PR so that we land on something solid

Reply:
Hey Ben, this is the release procedure that all features in PyTorch follow. We start with `prototype`, and then collect user feedbacks on 1) what's missing 2) what application will be benefit from this feature 3) etc. If no strong objection, we move on to `beta` to collect further feedback on, e.g., perf and also add more docs and tutorials. And finally, we land in `stable`, and after that BC-breaking changes should be avoided if possible.

Reply:
alright, sounds good, I was missing some context !